### ISSUE TITLE
Structure of package (NEPCore)

### ISSUE THREAD

**Issue by eringh** [nep-pack/NonlinearEigenproblems.jl#76]
We should probably think over how we structure the package.

Do you need to import `NEPCore` to be able to use the package?
What does the includes
```
using NonlinearEigenproblems.NEPSolver
using NonlinearEigenproblems.Gallery
```
mean in you do not have the `NEPCore`?

The question has to do with which modules are exporting what functions.
`NEPCore` is exporting the basic interface. Meaning that if you implement a specific part of the interface, it will exist and can be run. Even with out an `export`. Since you have overloaded an existing function.

My commit 3d1357e731ccd60626a54a59b7f098b8dd71add3 failed on one test. The error was `UndefVarError: compute_Mder not defined`. The problem was that I removed the export statements for that gallery-file. And since the test was having the `using`-statements above, there was no exported definition of `compute_Mder`. One fix, 4ece3de0068b458b4f7c84c23e716176a4a2ee4f, was just to make sure the `NEPCore`-module was loaded in the test.

So: Should we somehow redo the structure a bit and make sure that when the package is loaded, `NEPCore` is always loaded?
This might possibly mean that the code is moved out of the module and into the center of the package (of course the files does not have to be moved, just removing the module `NEPCore`).

**Comment by maxbennedich** [comment]
Sounds like a good idea to clean it up, although I don't know exactly what that would look like. It feels like at least NEPCore, NEPSolver, and NEPTypes could be included by a single `using` statement. Maybe more modules as well. My understanding is that using any of the modules will end up running `NonlinearEigenproblems.jl`, which will include all of the basic modules anyway, so performance-wise I don't think we have much to lose by loading them all?

**Comment by jarlebring** [comment]
I would not mind using Reexport if the only consequence is that we dont need the extra `using`. Someone wants to try it as PR? @eringh @maxbennedich

**Comment by eringh** [comment]
I can look at it.
What do we want to export? Is it "everything" as in solvers and types, or just "the core"?

**Comment by jarlebring** [comment]
Everything except gallery?

**Comment by maxbennedich** [comment]
My thinking is everything that end users might want to use, so that they only need to do `using NonlinearEigenproblems`. I'm wondering what the harm is in including more modules, such as Gallery and Serialization. As far as I can tell, it doesn't affect the initial load time, at least not as long as we're including these anyway in `NonlinearEigenproblems.jl`? (Or are we thinking of changing this?)

Another potential issue would be "namespace pollution", i.e. that we are making available a bunch of internal methods that users should never be calling and could cause confusion.

**Comment by eringh** [comment]
I was thinking of including everything in the `NonlinearEigenproblems.jl`.

An issue could be, as you mentioned, that we get namespace pollution.  However, it somehow feels like a user would like to load the package and then be done with it. In all the packages we use we do not load any extra modules from that package. We just write
```
    using LinearAlgebra
    using Polynomials
    using Random
```
to get the functionality.

Nevertheless, writing the code as 
```
    export NEPCore
    export NEPTypes
    export NEPSolver
```
will make the public code accessible through calls like `nep = NEPTypes.DEP([[1 2.0; 1 3]], [1.2])`. The user still need to explicitly import the required module, or call through the full names.

**Comment by eringh** [comment]
Fixed by #82
